Updated composite R2R format doc (#2221)
dotnet/runtime: docs/design/coreclr/botr/readytorun-format.md

Revisions:
* 1.1 - [Jan Kotas](https://github.com/jkotas) - 2015
* 3.1 - [Tomas Rylek](https://github.com/trylek) - 2019
* 4.1 - [Tomas Rylek](https://github.com/trylek) - 2020

# Introduction

This document describes ReadyToRun format 3.1 implemented in CoreCLR as of June 2019 and not yet
implemented proposed extensions 4.1 for the support of composite R2R file format. 
**Composite R2R file format** has basically the same structure as the traditional R2R file format
defined in earlier revisions except that the output file represents a larger number of input MSIL
assemblies compiled together as a logical unit.

# PE Headers and CLI Headers

**Single-file ReadyToRun images** conform to CLI file format as described in ECMA-335
with the following customizations:

- The PE file is always platform specific
- CLI Header Flags field has set `COMIMAGE_FLAGS_IL_LIBRARY` (0x00000004) bit set
- CLI Header `ManagedNativeHeader` points to READYTORUN_HEADER

The COR header and ECMA 335 metadata pointed to by the COM descriptor data directory item
in the COFF header represent a full copy of the input IL and MSIL metadata it was generated from.

> **Composite R2R files** currently conform to Windows PE executable file format as the

Reviewer: Eventually, we may have an option to have platform-native .so that have 1:1 mapping with IL images (I believe Mono has this option today). Do you expect that such image is still going to have the two layers of R2R headers?

Author: Hmm, frankly I was hoping that we wouldn't need yet another variation of the format (old-style MSIL  + R2R / native R2R with a single header / native R2R with two levels of headers) but it's certainly doable and probably not a huge concern as we'll need to continue supporting the single-level header for the old style case.

native envelope. Moving forward we plan to gradually add support for platform-native
executable formats (ELF on Linux, MachO on OSX) as the native envelopes. As a natural corollary

> There is no global CLI / COR header in the file. The ReadyToRun header structure is pointed to

Reviewer 2: `there` or the previous `As a natural corollary` is perhaps a truncated sentence.

Author: Right, "there" was supposed to be part of the same sentence. Thanks for spotting, fixed in 3rd commit.

Revisions:
* 1.1 - [Jan Kotas](https://github.com/jkotas) - 2015
* 3.1 - [Tomas Rylek](https://github.com/trylek) - 2019

> * 3.2 - [Tomas Rylek](https://github.com/trylek) - 2020

Author: Fixed in 2nd commit, thanks for pointing that out!

# Introduction

This document describes ReadyToRun format 3.1 implemented in CoreCLR as of June 2019 and not yet
implemented proposed extensions 3.2 for the support of composite R2R file format. 
**Composite R2R file format** has basically the same structure as the traditional R2R file format
defined in earlier revisions except that the output file represents a larger number of input MSIL
assemblies compiled together as a logical unit.

> **Note**: The addition of the new composite R2R file format flavor doesn't require bumping up the

Author: I have removed the **Note** paragraph in 2nd commit.

major ReadyToRun format version number. This is because, according to the format definition, the
composite R2R file format doesn't technically conform to the single-input R2R supported by older
versions of CoreCLR so there's no risk of *"the old loader"* messing up by incorrectly trying to
run *"the new file"*. The only downside is that the *"new composite R2R file"* won't run with the
*"old loader"* and thus it will somewhat violate the design principle that R2R is a mere code cache.

# PE Headers and CLI Headers

**Single-file ReadyToRun images** conform to CLI file format as described in ECMA-335
with the following customizations:

- The PE file is always platform specific
- CLI Header Flags field has set `COMIMAGE_FLAGS_IL_LIBRARY` (0x00000004) bit set
- CLI Header `ManagedNativeHeader` points to READYTORUN_HEADER

The COR header and ECMA 335 metadata pointed to by the COM descriptor data directory item
in the COFF header represent a full copy of the input IL and MSIL metadata it was generated from.

> For **Composite R2R files** there is no global CLI / COR header (as there are potentially

Author: I tried to reformulate this in 2nd commit.

# Structures

The structures and accompanying constants are defined in the [readytorun.h]
(https://github.com/dotnet/runtime/blob/master/src/coreclr/src/inc/readytorun.h) header file.
Basically the entire R2R executable image is addressed through the READYTORUN_HEADER singleton
pointed to by the well-known export RTR_HEADER in the export section of the native executable
envelope.

> For single-file R2R executables, there's just one header representing all image sections.

Reviewer 2: Seems a bit confusing, as far as I can tell there's only one header no matter if the file is composite or not.

Author: @mikedn - I must admit I'm somewhat unsure what you're trying to say: it seems to me you're either trying to say that CoreCLR runtime and Crossgen1/2 always only generate one header - that is correct, this is a feature proposal that requires adding the support for per-assembly sub-headers - or that the proposal is poorly written in the sense of insufficiently or inconsistently referring to header levels. Could you please clarify?

Based on how I'm imagining the design and based on JanK's comments the version I'm about to publish as the 3rd commit on this review distinguishes three different cases:

1) Legacy case - Windows MSIL PE with attached R2R data (single-assembly only);

2) Modern single-assembly R2R - initially Windows native PE, ultimately OS-native executable format, single R2R header locatable via well-known export;

3) Modern multiple-assembly composite / single-exe R2R - initially Windows native PE, ultimately OS-native executable format, master R2R header locatable via well-known export plus assembly table pointing to per assembly subheaders with assembly-level sections (e.g. AVAILABLE TYPES).

Reviewer 2: > or that the proposal is poorly written in the sense of insufficiently or inconsistently referring to header levels. Could you please clarify?

It's a question of phrasing. Don't both single and composite files have a single `READYTORUN_HEADER` and the difference between them is the presence of a `READYTORUN_SECTION_ASSEMBLIES` section?

Author: That is true but the main purpose of the ASSEMBLIES section is to encode a series of per assembly headers (of the type <code>READYTORUN_CORE_HEADER</code>) that have a similar structure and purpose as the main R2R header - to encode a set of sections specific to a particular assembly. I'll take another look to see how to make it more obvious; I have also slightly reformulated this part in 3rd commit based on JanK's comments.

Reviewer 2: Thanks, I think simply specifying the header name would be sufficient. Something like:

> For single-file R2R executables, there's just one `READYTORUN_CORE_HEADER` header representing all image sections.

In the composite case the text already says:

> Each `READYTORUN_SECTION_ASSEMBLIES_ENTRY` record points  to a `READYTORUN_CORE_HEADER` variable-length structure representing sections specific to the particular assembly.

so it should be clear enough.

Author: I have noticed that I missed updating the description of the ASSEMBLIES section. This should be fixed now as of 4th commit, I'll certainly appreciate if you could take another look to make sure the description is fully consistent.

successfully execute native code from images of version 2.9. The runtime should refuse to execute
native code from image of version 3.0.

## READYTORUN_CORE_HEADER

```C++
struct READYTORUN_CORE_HEADER
{
    DWORD                   Flags;          // READYTORUN_FLAG_XXX

    DWORD                   NumberOfSections;

    // Array of sections follows. The array entries are sorted by Type
    // READYTORUN_SECTION   Sections[];
};
```

### READYTORUN_CORE_HEADER::Flags

| Flag                                    |      Value | Description
|:----------------------------------------|-----------:|:-----------
| READYTORUN_FLAG_PLATFORM_NEUTRAL_SOURCE | 0x00000001 | Set if the original IL image was platform neutral. The platform neutrality is part of assembly name. This flag can be used to reconstruct the full original assembly name.
| READYTORUN_FLAG_COMPOSITE               | 0x00000002 | The image represents a composite R2R file resulting from a combined compilation of a larger number of input MSIL assemblies.
| READYTORUN_FLAG_EMBEDDED_MSIL           | 0x00000004 | Input MSIL is embedded in the R2R image.

> | READYTORUN_FLAG_COMPONENT               | 0x00000008 | This is 

Reviewer: Unfinished sentence?

Author: Thanks for spotting, fixed in 3rd commit.

|         0 | defined type
|         1 | exported type

The version-resilient hashing algorithm used for hashing the type names is implemented in
[vm/versionresilienthashcode.cpp](https://github.com/dotnet/runtime/blob/8c6b1314c95857b9e2f5c222a10f2f089ee02dfe/src/coreclr/src/vm/versionresilienthashcode.cpp#L75).

**Note:** This is a per-assembly section. In single-file R2R files, it is pointed to directly by the
main R2R header; in composite R2R files, each component module has its own available type section pointed to
by the `READYTORUN_SECTION_ASSEMBLIES_ENTRY` core header structure.

## READYTORUN_SECTION_INSTANCE_METHOD_ENTRYPOINTS

This section contains a native hashtable of all generic method instantiations compiled into
the R2R executable. The key is the method instance signature; the appropriate version-resilient
hash code calculation is implemented in
[vm/versionresilienthashcode.cpp](https://github.com/dotnet/runtime/blob/master/src/coreclr/src/vm/versionresilienthashcode.cpp#L127);
the value, represented by the `EntryPointWithBlobVertex` class, stores the method index in the
runtime function table, the fixups blob and a blob encoding the method signature.

> **Note:** In contrast to non-generic method entrypoints, this section is image-wide for

Reviewer: > As mentioned elsewhere in this document, CoreCLR runtime requires changes to
properly look up methods stored in this section in the composite R2R case.

This should be in design notes.

Author: Good point, moved to design notes in 3rd commit.

## READYTORUN_SECTION_MANIFEST_METADATA

Manifest metadata is an [ECMA-335] metadata blob containing extra reference assemblies within
the version bubble introduced by inlining on top of assembly references stored in the input MSIL.
As of R2R version 3.1, the metadata is only searched for the AssemblyRef table. This is used to
translate module override indices in signatures to the actual reference modules (using either
the `READYTORUN_FIXUP_ModuleOverride` bit flag on the signature fixup byte or the
`ELEMENT_TYPE_MODULE_ZAPSIG` COR element type).

> **Disclaimer:** The manifest metadata is a new feature that hasn't shipped yet; it involves

Reviewer: The note that this evolved from fragile NGen can be in design notes. It should not be in the file format spec. Many other things in the file format spec evolved from fragile NGen, but we do not mention it anywhere.

Author: Moved to design notes in 2nd commit.

## READYTORUN_SECTION_MANIFEST_METADATA

Manifest metadata is an [ECMA-335] metadata blob containing extra reference assemblies within
the version bubble introduced by inlining on top of assembly references stored in the input MSIL.

> As of R2R version 3.1, the metadata is only searched for the AssemblyRef table. This is used to

Reviewer: ```suggestion
As of R2R version 3.1, the metadata is only used for the AssemblyRef table. This is used to
```

Author: Fixed in 3rd commit.

**TODO**: document attribute presence encoding

**Note**: We already know this table uses assembly-relative token encoding so it has similar
characteristics like `READYTORUN_SECTION_AVAILABLE_TYPES` or `READYTORUN_SECTION_METHOD_ENTRYPOINTS`.
No matter what component assembly-relative encoding we end up choosing for these tables, we
should use the same encoding for ATTRIBUTEPRESENCE.

## READYTORUN_SECTION_ASSEMBLIES (v3.2+)

> This section is only present in composite R2R files. It is a straight binary array of the

Reviewer: ```suggestion
This section is only present in composite R2R files. It is an array of the
```

Author: Fixed in 2nd commit.

entries `READYTORUN_SECTION_ASSEMBLIES_ENTRY` parallel to the indices in the manifest metadata
AssemblyRef table in the sense that it's a linear table where the row indices correspond to the
equivalent AssemblyRef indices. Just like in the AssemblyRef ECMA 335 table, the indexing is
1-based (the first entry in the table corresponds to index 1).

```C++

> struct READYTORUN_SECTION_ASSEMBLIES_ENTRY

Reviewer 3: This shouldn't be a hard-coded list of items what are assembly local. We probably need this to be a variable sized structure, described by section entries just like the outer section concept.

Author: I think you're right, I've also been thinking about this when resurrecting my doc from last summer. I'll update the document based on your suggestion.

Author: The only trick is how to pull this off without introducing an unintentional <code>O(n)</code> lookup when searching for, say, the per-module entrypoint table - but you may argue that <code>n</code> is usually very small (less than ten with a huge headroom).

Reviewer 3: Mmm.. possibly. I rather imagine it won't really be an O(n) loop at lookup time, more of a thing where we capture all of the interesting data into a structure which is of fixed size and shape in the assembly loader. For instance, we effectively have that loop now in R2R, and I've never seen it have a high cost that matters anywhere.

Author: OK, that sounds reasonable, even Bing with its 1K-ish number of assemblies shouldn't pose a big problem for parsing a simple runtime structure and filling in some internal array of structures in the CoreCLR runtime.

Author: In 2nd commit I have changed the proposed design so that the R2R header is basically two-level: for composite images, the master header includes the section <code>READYTORUN_SECTION_ASSEMBLIES</code> that for each component assembly points at its own sub-header with a list of sections.

Reviewer 4: @trylek If that number were 10K would it still be reasonable to have it this way? I'm not suggesting 10K be a benchmark but it would be nice to either know this is a one-time startup cost and therefore it is ok to pay that for simplicity.

Author: @mjsabby My personal feeling is that, even with 10K assemblies, other costs still far outweigh parsing the headers and storing the information somewhere; in general you're right however, the easiest way to improve asymptotic startup is by minimizing any forms of runtime preprocessing.

**TODO**: document attribute presence encoding

**Note:** This is a per-assembly section. In single-file R2R files, it is pointed to directly by the
main R2R header; in composite R2R files, each component module has its own attribute presence
section pointed to by the `READYTORUN_SECTION_ASSEMBLIES_ENTRY` core header structure.

## READYTORUN_SECTION_ASSEMBLIES (v3.2+)

This image-wide section is only present in the main R2R header of composite R2R files. It is an
array of the entries `READYTORUN_SECTION_ASSEMBLIES_ENTRY` parallel to the indices in the manifest metadata
AssemblyRef table in the sense that it's a linear table where the row indices correspond to the
equivalent AssemblyRef indices. Just like in the AssemblyRef ECMA 335 table, the indexing is
1-based (the first entry in the table corresponds to index 1).

```C++
struct READYTORUN_SECTION_ASSEMBLIES_ENTRY
{
    IMAGE_DATA_DIRECTORY CorHeader;         // Input MSIL metadata COR header
    IMAGE_DATA_DIRECTORY AvailableTypes;    // Available types table
    IMAGE_DATA_DIRECTORY MethodEntrypoints; // Method entrypoint table

> };

Reviewer 2: And speaking of

> Each READYTORUN_SECTION_ASSEMBLIES_ENTRY record points to a READYTORUN_CORE_HEADER variable-length structure representing sections specific to the particular assembly.

Where exactly is `READYTORUN_CORE_HEADER`?

Author: Around line 95 in the readytorun-format.md file.

Reviewer 2: No, I was asking where the record "points to" that header. I see that you have updated this structure in the last commit so it looks good now. Thanks!

**TODO**: document attribute presence encoding

> **Note**: We already know this table uses assembly-relative token encoding so it has similar

Reviewer: I am not sure what this is trying to say. Should this just say: "**Note:** This is a per-assembly section. ... " ?

Author: I guess I was trying to say "I'm not sure" at the time of writing the document. As you confirm my suspicion that this is a per-assembly table, I have changed it as you suggested in the 3rd commit.

characteristics like `READYTORUN_SECTION_AVAILABLE_TYPES` or `READYTORUN_SECTION_METHOD_ENTRYPOINTS`.
No matter what component assembly-relative encoding we end up choosing for these tables, we
should use the same encoding for ATTRIBUTEPRESENCE.

> ## READYTORUN_SECTION_ASSEMBLIES (v3.2+)

Reviewer 2: A table above states a different version:

> Image (added in V4.1)

Author: Fixed in 4th iteration.
